### ISSUE TITLE
When useJarUrls is configured for classloading virtual directory URLs should not be returned from the app class loader

### ISSUE THREAD

**Issue by tjwatson** [OpenLiberty/open-liberty#28405]
When deploying a Spring Boot application that contains a dependency JAR which has classes contained in sub-packages that should be scanned (see https://www.baeldung.com/spring-component-scanning) if the dependency JAR does not contain directory entries then something like the following exception will occur when starting the Spring Boot application:

```
org.springframework.beans.factory.BeanDefinitionStoreException: I/O failure during classpath scanning
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.scanCandidateComponents(ClassPathScanningCandidateComponentProvider.java:515) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.findCandidateComponents(ClassPathScanningCandidateComponentProvider.java:351) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ClassPathBeanDefinitionScanner.doScan(ClassPathBeanDefinitionScanner.java:277) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ComponentScanAnnotationParser.parse(ComponentScanAnnotationParser.java:128) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ConfigurationClassParser.doProcessConfigurationClass(ConfigurationClassParser.java:306) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ConfigurationClassParser.processConfigurationClass(ConfigurationClassParser.java:246) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:197) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:165) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:416) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:289) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:349) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:118) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:788) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:606) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.2.5.jar:3.2.5]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754) ~[spring-boot-3.2.5.jar:3.2.5]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456) ~[spring-boot-3.2.5.jar:3.2.5]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:334) ~[spring-boot-3.2.5.jar:3.2.5]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1354) ~[spring-boot-3.2.5.jar:3.2.5]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1343) ~[spring-boot-3.2.5.jar:3.2.5]
	at com.example.demo.DemoApplication.main(DemoApplication.java:10) ~[demo-0.0.1-SNAPSHOT.spring:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:586) ~[na:na]
	at com.ibm.ws.app.manager.springboot.internal.SpringBootRuntimeContainer.lambda$invokeSpringMain$6(SpringBootRuntimeContainer.java:139) ~[na:na]
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:280) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:1595) ~[na:na]
Caused by: java.io.FileNotFoundException: JAR entry com/example/demo/ not found in /home/tjwatson/dev/liberty/source/open-liberty/dev/build.image/wlp/usr/servers/spring/workarea/spring/lib.index.cache/72/db0fb4f1c2c78a6adc102681fe49409ab019b07413923f0bea6343c7638a2b/demo.dependency-1.0.0.jar
	at java.base/sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:124) ~[na:na]
	at java.base/sun.net.www.protocol.jar.JarURLConnection.getJarFile(JarURLConnection.java:70) ~[na:na]
	at org.springframework.core.io.support.PathMatchingResourcePatternResolver.doFindPathMatchingJarResources(PathMatchingResourcePatternResolver.java:683) ~[spring-core-6.1.6.jar:6.1.6]
	at org.springframework.core.io.support.PathMatchingResourcePatternResolver.findPathMatchingResources(PathMatchingResourcePatternResolver.java:586) ~[spring-core-6.1.6.jar:6.1.6]
	at org.springframework.core.io.support.PathMatchingResourcePatternResolver.getResources(PathMatchingResourcePatternResolver.java:334) ~[spring-core-6.1.6.jar:6.1.6]
	at org.springframework.context.support.AbstractApplicationContext.getResources(AbstractApplicationContext.java:1511) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.support.GenericApplicationContext.getResources(GenericApplicationContext.java:262) ~[spring-context-6.1.6.jar:6.1.6]
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.scanCandidateComponents(ClassPathScanningCandidateComponentProvider.java:457) ~[spring-context-6.1.6.jar:6.1.6]
```

In the above example all class path entries (application code and its dependency JARs) should be scanned for classes contained in the package `com.example.demo` and any sub-packages of `com.example.demo`.  This exception occurred because the dependency JAR `demo.dependency-1.0.0.jar` contained classes in a sub-package of `com.example.demo` but it contained no directory entries at all.

The problem here is that the Liberty class loader method `getResources` is returning URLs when a call is made for the resource path `com/example/demo` from this `demo.dependency-1.0.0.jar` even though there is no JAR entry for that directory path.  Here there are classes in a sub-package (e.g. `com/example/demo/util/Test.class`), but none of the directory entries leading up to the resource for the classes do exist.  Liberty's `ZipFileContainer` detects this and will return entries that are virtualized for the the directory paths because it knows entry files do exist in sub-folders of the requested directory path.

When the `jar` protocol is used then Spring will attempt to take the `jar` URL and call `openConnection` to get a `JarURLConnection` and then it attempts call `getJarFile` on the `JarURLConnection` which fails because it cannot connect to the non-existing entry in the JAR file.

Can we avoid returning URLs when `useJarUrls` is configured for "virtualized" entries in the JAR file that do not exist?

Possible Options:

1. Do nothing, force applications to have properly formed JARs with the necessary directory entries that Spring expects.  This has a disadvantage with providing parity for running Spring Boot applications on Liberty.  Such applications will start, without error, when launched outside of liberty (e.g. using `java -jar app.jar`).  But in such cases the dependency JAR will be skipped for class path scanning, even though it really should be scanned.
2. If `useJarUrls` is configured for the application class loader then do not have the class loader return `jar` URLs to entries that do not exist in the JAR/ZIP file.  This will provide parity with how Spring Boot applications behave outside of Liberty, but has a risk that it may break others that depend on the `jar` URLs being returned.  But such a scenario would never actually try to connect to the entry because they would face the same error.  This risk may be very low.
3. Similar to option 2, but only do it for class loaders that are used for an application deployed with one of the `springBoot-x.x` features.  This helps minimize the risk of 2, but does not solve the problem for when the spring boot application is deployed as a `webApplication` and `useJarUrls` is configured.
4. Similar to option 2, but introduce another config option to class loader that allows configuring the classloader to not return virtual `jar` protocol URLs.  Minimizes risk if the default keeps the current behavior.  Documenting and naming such an option could be difficult and confusing.

Currently my preferred option is 2, followed by 1.  I don't really want to do 3 or 4.

**Comment by tjwatson** [comment]
The preferred option on the design call was option 2, but with a new warning that informs the user that we detected the situation where they have a dependency JAR without directory entries and the result is that Spring will not scan the JAR correctly for classes that do exist in the JAR in sub-packages that should be scanned.

The error message will need to be generalized though since it could apply to any application where `userJarUrls` was configured.  Something like the following:

> The JAR {0} does not have directory entries for classes and/or resources contained in sub-folders of the {1} directory. This may cause some frameworks to ignore the JAR for scanning purposes.  For example, the Spring framework will not scan this JAR.

**Comment by tjwatson** [comment]
Fixed with #28404
